src/mock/v2.0/intent-k.ts

```diff
@@ -15,6 +15,7 @@ onFdc3Ready().then(async () => {
     const privChan = await fdc3.createPrivateChannel();
 
     await privChan.addContextListener(ContextType.testContextX, async () => {
+      await wait(100); //wait for listener in test to initialise
```

Comment:
this should not be necessary - the listener should be set-up before reaching this part of the test. Doing it this way is likely to be fragile.

Reply:
Ah I see the architectural problem that led to this.  In the test definition here: https://github.com/finos/FDC3-conformance-framework/blob/2e42c642a7a020075b71d18f8b7a0665790e7103/src/test/v2.0/advanced/fdc3.raiseIntent.ts#L103-L129
The listeners being added via `receiveContext` is not awaited (because it can't be) and hence it doesn't happen without a sleep (javascript keeps running synchronously until it hits some other async call - then runs the various async calls). It can't be awaited as `receiveContext` is poorly designed: https://github.com/finos/FDC3-conformance-framework/blob/59884d2e1ace986bcf6a5fceda634573adb506e0/src/test/v2.0/support/intent-support-2.0.ts#L12-L28

FDC3 API calls are async - although `receiveContext` awaits `addContextListener` you can't await `receiveContext` as it only resolves when the listener receives something - i.e. its got two different async events in one function. The correct solution to this would have been to have `receiveContext`  return an object with the listener's promise embedded in it. You could then await `receiveContext` and be sure the listener was added and receive the object back. You can then separately await the second promise that will resolve when the listener receives data.

This needs fixing properly as it's going to be introducing races and fragility everywhere `receiveContext`is used. Sounds hard, but isn't really. Once understood it should only take a few mins to refactor all uses of `receiveContext`. The thing to remember is that you should not mix `new Promise` with `await` in the same function - it's a well known anti-pattern. 

`receiveContext` should change to:

```ts
async receiveContext(contextType: string, waitTime?: number): Promise<AppControlContext> {
    let timeout;
    const appControlChannel = await getOrCreateChannel(constants.ControlChannel);
    //wrap promise so we can await this function without it having to be resolved
    return { listenerPromise: new Promise<Context>(async (resolve, reject) => {
      const listener = await appControlChannel.addContextListener(contextType, (context: AppControlContext) => {
        resolve(context);
        clearTimeout(timeout);
        listener.unsubscribe();
      });

      //if no context received reject promise
      const { promise: sleepPromise, timeout: theTimeout } = sleep(waitTime ?? constants.WaitTime);
      timeout = theTimeout;
      await sleepPromise;
      reject(new Error("No context received. Listener expected to receive context of type " + contextType + " from mock app"));
    })};
  }
```

Then tests using it (e.g. [this one](https://github.com/finos/FDC3-conformance-framework/blob/2e42c642a7a020075b71d18f8b7a0665790e7103/src/test/v2.0/advanced/fdc3.raiseIntent.ts#L103-L129) should change calls from:

```ts
      let textContextXReceiver = control.receiveContext(ContextType.testContextX);
      control.privateChannelBroadcast(<PrivateChannel>result, ContextType.testContextX);
      await textContextXReceiver;
```

to:
```ts
      let { listenerPromise: textContextXReceiver } = await control.receiveContext(ContextType.testContextX);
      control.privateChannelBroadcast(<PrivateChannel>result, ContextType.testContextX);
      await textContextXReceiver;
```

Then no fragile waits are needed.

If you need a temporary fix, move the wait from here into the test before the `privateChannelBroadcast`... but I'd recommend the quick refactor as it won't take long and this is a horrible example of an antipattern.